internal/fans/nvidia.go

```diff
@@ -0,0 +1,228 @@
+	pwm = min(pwm, 100)
+	ret := device.SetFanSpeed_v2(fanIdx, pwm)
+	return nvError(ret)
+}
+
+func (fan *NvidiaFan) GetFanRpmCurveData() *map[int]float64 {
+	return fan.FanCurveData
+}
+
+// AttachFanCurveData attaches fan curve data from persistence to a fan
+// Note: When the given data is incomplete, all values up until the highest
+// value in the given dataset will be interpolated linearly
+// returns os.ErrInvalid if curveData is void of any data
+func (fan *NvidiaFan) AttachFanRpmCurveData(curveData *map[int]float64) (err error) {
+	if curveData == nil || len(*curveData) <= 0 {
+		ui.Error("Cant attach empty fan curve data to fan %s", fan.GetId())
+		return os.ErrInvalid
+	}
+
+	fan.FanCurveData = curveData
+
+	startPwm, maxPwm := ComputePwmBoundaries(fan)
+	fan.SetStartPwm(startPwm, false)
+	fan.SetMaxPwm(maxPwm, false)
+
+	// TODO: we don't have a way to determine this yet
+	fan.SetMinPwm(startPwm, false)
+
+	return err
+}
+
+func (fan *NvidiaFan) UpdateFanRpmCurveValue(pwm int, rpm float64) {
+	if fan.FanCurveData == nil {
+		fan.FanCurveData = &map[int]float64{}
+	}
+	(*fan.FanCurveData)[pwm] = rpm
+}
+
+func (fan *NvidiaFan) GetCurveId() string {
+	return fan.Config.Curve
+}
+
+func (fan *NvidiaFan) ShouldNeverStop() bool {
+	return fan.Config.NeverStop
+}
+
+func (fan *NvidiaFan) GetPwmEnabled() (int, error) {
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	policy, err := device.GetFanControlPolicy_v2(fanIdx)
+	if err != nvml.SUCCESS {
+		return 2, nvError(err)
+	}
+	pwm := 2 // "motherboard pwm control" as default assumption
+	if policy == nvml.FAN_POLICY_MANUAL {
+		if fan.RunAtMaxSpeed {
+			pwm = 0 // "max speed" - "no control" in hwmon backend
+		} else {
+			pwm = 1 // manual PWM control
+		}
+	}
+	return pwm, nil
+}
+
+func (fan *NvidiaFan) IsPwmAuto() (bool, error) {
+	value, err := fan.GetPwmEnabled()
+	if err != nil {
+		return true, err // assume auto control by default
+	}
+	return value == 2, nil
+}
+
+// SetPwmEnabled writes the given value to pwmX_enable
+// Possible values (unsure if these are true for all scenarios):
+// 0 - no control (results in max speed)
+// 1 - manual pwm control
+// 2 - motherboard pwm control
+func (fan *NvidiaFan) SetPwmEnabled(value ControlMode) (err error) {
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	var ret nvml.Return = nvml.SUCCESS
+	if value == 2 {
+		ret = nvml.DeviceSetDefaultFanSpeed_v2(device, fanIdx)
+	} else {
+		// TODO: really support mode 0? here mode 2 is default and mode 0 must be emulated
+		if value == 0 {
+			fan.RunAtMaxSpeed = true
+			ret = nvml.DeviceSetFanSpeed_v2(device, fanIdx, 100)
+		} else {
+			ret = device.SetFanControlPolicy(fanIdx, nvml.FAN_POLICY_MANUAL)
+			// TODO: set speed? just setting a speed implicitly sets manual policy - if so, what speed?
+		}
+	}
+	return nvError(ret)
+}
+
+func (fan *NvidiaFan) Supports(feature FeatureFlag) bool {
+	return (fan.SupportedFeatures & (1 << feature)) != 0
```

Comment:
Is it necessary to remember the `SupportedFeatures` in the struct?
If there is not a huge penalty, I would like to avoid this, in case it changes during operation.
Also, although nifty, I am not a fan of this notation, as it is not very readable IMHO.
Handling each feature flag in a `switch` was a deliberate choice in the hwmon implementation.

Reply:
> How should this change during operation?

Well, I have no idea, but oh boy let me tell you... fan control is not as simple as I though when I started this project :laughing: 
Assume nothing, expect everything, and you will still get surprised... :sweat_smile: 

But on topic: I would like to have this method actually implement the check that determines whether the feature is available or not. Essentially the code that you already wrote in the detection stuff. If we need to cache this to avoid performance issues, thats fine, but I would like to have this logic available in the Fan implementation to be able to recheck if "properly" if necessary.

Reply:
> but I would like to have this logic available in the Fan implementation

This would mean kind of duplicating at least parts of the fan detection code between nvidia/nvidia.go and fans/nvidia.go

However, I have rough ideas for backend refactoring (as mentioned yesterday), I'll have to think it through a bit more and then document it somewhere so we can discuss it, but it would make this kind of thing easier and wouldn't probably also prevent duplicating code.

Reply:
I turned that integer with bit flags into three bools now.
They're set in NvidiaFan.Init()